### ISSUE TITLE
python GhostDriver fails in send_keys to GitHub signin

### ISSUE THREAD

**Issue by tlee5794** [detro/ghostdriver#445]
Trying to automate a GitHub login with PhantomJS 2.0 with the following python script snippet:

```
class GitHubLogIn(unittest.TestCase):
    def setUp(self):
        self.driver = webdriver.PhantomJS('./phantomjs/bin/phantomjs')
        self.driver.set_window_size(1000, 1000)
        self.driver.implicitly_wait(30)
        self.base_url = "https://github.com/"
        self.verificationErrors = []
        self.accept_next_alert = True
    
    def test_git_hub_log_in(self):
        driver = self.driver
        driver.get(self.base_url + "/")
        driver.find_element_by_link_text("Sign in").click()
        time.sleep(10)
        element = driver.find_element_by_id("login_field")
        element.click()
        element.clear()
        element.send_keys("username") 
```

The send_keys fails with:

```
======================================================================
ERROR: test_git_hub_log_in (__main__.GitHubLogIn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "scripts\Test4.py", line 72, in test_git_hub_log_in
    element.send_keys("username")
  File "C:\Python27\lib\site-packages\selenium\webdriver\remote\webelement.py", line 318, in send_keys
    self._execute(Command.SEND_KEYS_TO_ELEMENT, {'value': typing})
  File "C:\Python27\lib\site-packages\selenium\webdriver\remote\webelement.py", line 402, in _execute
    return self._parent.execute(command, params)
  File "C:\Python27\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 175, in execute
    self.error_handler.check_response(response)
  File "C:\Python27\lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 136, in check_response
    raise exception_class(value)
WebDriverException: Message: TypeError - undefined is not a constructor (evaluating '_getTagName(currWindow).toLowerCase()')
```

The send_keys works fine with the Chrome or the Firefox WebDriver. Any ideas how to get the send_keys above to work with GhostDriver or what the possible issue would be?

**Comment by asacamano** [comment]
Hi,

I'm seeing the same thing... 

Did you ever come up with  solution?

Thanks,

Andrew

**Comment by asacamano** [comment]
And I was able to solve my problem by being more specific about the conditions I was waiting for when I got the element - see https://github.com/detro/ghostdriver/issues/450#issuecomment-111250483.

It seems to me to be related to calling into ghostdriver when the element is not entirely constructed.

**Comment by mirraj2** [comment]
I have this problem as well with the site I am automating.  I waited ten seconds after the site was loaded AND tested to make sure that the username field is displayed and enabled.  Waiting did not work for me.  There is some other problem.

org.openqa.selenium.UnsupportedCommandException: TypeError - undefined is not a constructor (evaluating '_getTagName(currWindow).toLowerCase()')
Command duration or timeout: 6 milliseconds
Build info: version: '2.45.0', revision: '32a636c', time: '2015-03-05 22:01:35'
System info: host: 'Jasons-MacBook-Pro.local', ip: '10.0.1.9', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.10.3', java.version: '1.8.0_45'
Driver info: org.openqa.selenium.phantomjs.PhantomJSDriver
Capabilities [{applicationCacheEnabled=false, rotatable=false, handlesAlerts=false, databaseEnabled=false, version=2.0.0, platform=OS X 10.9, browserConnectionEnabled=false, proxy={proxyType=direct}, nativeEvents=true, acceptSslCerts=false, driverVersion=1.2.0, phantomjs.page.settings.userAgent=Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2228.0 Safari/537.36, locationContextEnabled=false, webStorageEnabled=false, browserName=phantomjs, takesScreenshot=true, driverName=ghostdriver, javascriptEnabled=true, cssSelectorsEnabled=true}]
Session ID: 56ce1e90-2066-11e5-925b-7b8513fcbb41
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:204)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:156)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:599)
	at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:268)
	at org.openqa.selenium.remote.RemoteWebElement.sendKeys(RemoteWebElement.java:89)

**Comment by clickthemouse** [comment]
@mirraj2 : I have a similar error (windows 7 python 2.7)

selenium.common.exceptions.WebDriverException: Message: TypeError - undefined is
 not a constructor (evaluating '_getTagName(currWindow).toLowerCase()')

Waiting don't work at all, I used save_screenshot and it seems it just don't work to work. Help!

**Comment by clickthemouse** [comment]
@asacamano : Can you help me I can't get it work!
Post a script where it works and where it don't works, personally I use python, the same error but waiting don't work :/
HELPPPPPPP MATE! :)

**Comment by Legioth** [comment]
I had a similar issue together with lots of StaleElementReferenceException from Selenium. It turned out that the Shadow DOM polyfill from http://webcomponents.org/polyfills/ changes some DOM apis so that GhostDriver (or Selenium?) fails to properly keep track of elements.

I resolved the issue by changing the page to use webcomponents-lite.js instead.

**Comment by fotisp** [comment]
@Legioth what if the site in question is not mine? This is a general issue which I am also having and I can't ask the website I am testing against to change **their** code, so that **my** ghostdriver code works.

- I am using java 1.8
- Waiting is not the issue
- sendKeys on any element will crash with the following exception

``` Java
    public String[] getRecords(){

        Capabilities caps = new DesiredCapabilities();
        ((DesiredCapabilities) caps).setJavascriptEnabled(true);
        ((DesiredCapabilities) caps).setCapability("takesScreenshot", true);
        ((DesiredCapabilities) caps).setCapability(
                PhantomJSDriverService.PHANTOMJS_EXECUTABLE_PATH_PROPERTY,
                applicationSettings.system.phantom
        )

        WebDriver driver = new PhantomJSDriver(caps)
        driver.get("*****")

        WebDriverWait wait = new WebDriverWait(driver, 5);
        wait.until(presenceOfElementLocated(By.className("IframeMainFormContent")));

        wait = new WebDriverWait(driver, 5);
        wait.until(presenceOfElementLocated(By.id("Username")));

        wait = new WebDriverWait(driver, 5);
        wait.until(presenceOfElementLocated(By.id("Password")));

        logger.debug("Form element is available, moving on")
        FileUtils.copyFile(driver.getScreenshotAs(OutputType.FILE),new File("/tmp/on-load.png"))

        WebElement element =driver.findElement(By.className("IframeMainFormContent"))
        element.findElement(By.id("Username")).sendKeys("123456789")
        element.findElement(By.id("Password")).sendKeys("supersecret")

        FileUtils.copyFile(driver.getScreenshotAs(OutputType.FILE),new File("/tmp/before-login.png"))

        element.submit()

        FileUtils.copyFile(driver.getScreenshotAs(OutputType.FILE),new File("/tmp/after-login.png"))
        driver.quit()
        return []
    }

    private static Function<WebDriver,WebElement> presenceOfElementLocated(final By locator) {
        return new Function<WebDriver, WebElement>() {
            @Override
            public WebElement apply(WebDriver driver) {
                return driver.findElement(locator);
            }
        };
    }

```
```
 [ERROR - 2015-09-01T12:54:03.764Z] RouterReqHand - _handle.error - {"line":264,"sourceURL":":/ghostdriver/request_handlers/webelement_request_handler.js","stack":"\tat _postValueCommand (:/ghostdriver/request_handlers/webelement_request_handler.js:264:52)\n\tat _handle (:/ghostdriver/request_handlers/webelement_request_handler.js:72:30)\n\tat _reroute (:/ghostdriver/request_handlers/request_handler.js:61:20)\n\tat _handle (:/ghostdriver/request_handlers/session_request_handler.js:120:42)\n\tat _reroute (:/ghostdriver/request_handlers/request_handler.js:61:20)\n\tat _handle (:/ghostdriver/request_handlers/router_request_handler.js:78:46)"}

  :262 in error
Exception in thread "main" org.openqa.selenium.UnsupportedCommandException: TypeError - undefined is not a constructor (evaluating '_getTagName(currWindow).toLowerCase()')
Command duration or timeout: 9 milliseconds
Build info: version: '2.47.1', revision: 'unknown', time: '2015-07-30 11:02:44'
System info: host: '***', ip: '***', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.10.4', java.version: '1.8.0_45'
Driver info: org.openqa.selenium.phantomjs.PhantomJSDriver
Capabilities [{applicationCacheEnabled=false, rotatable=false, handlesAlerts=false, databaseEnabled=false, version=2.0.0, platform=OS X 10.9, browserConnectionEnabled=false, proxy={proxyType=direct}, nativeEvents=true, acceptSslCerts=false, driverVersion=1.2.0, locationContextEnabled=false, webStorageEnabled=false, browserName=phantomjs, takesScreenshot=true, driverName=ghostdriver, javascriptEnabled=true, cssSelectorsEnabled=true}]
Session ID: 8553af10-50a8-11e5-a2b4-a7f3f744ed05
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:595)
	at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:273)
	at org.openqa.selenium.remote.RemoteWebElement.sendKeys(RemoteWebElement.java:94)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoCachedMethodSite.invoke(PojoMetaMethodSite.java:189)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:53)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:120)

```

**Comment by reservoirdog** [comment]
Are there any updates on this bug? I have the same problem.
The issue doesn't occur when I use v1.9.8 (the test case actually works and sendKeys doesn't crash)

**Comment by detro** [comment]
I don't know why no one bothers to read the README of the project: no one worked on porting this to PhatomJS 2.0.
It was imported, but no one actually took care of checking all the functionalities was still intact.

Do not use PhantomJS 2.0 if you want to use GhostDriver (i.e. webdriver mode).
Otherwise, you are on your own.

**Comment by turtleDev** [comment]
I still have this issue with PhantomJS 1.9.8. Maybe we should reopen this issue ?

**Comment by monkpit** [comment]
I don't see that anywhere in the readme. What is the cause of the issue and what is the resolution?

**Comment by khebbie** [comment]
Hello!

 

You have a new message, please read <http://kuyubasifidanligi.com/high.php?dy>

 

klaus@hebsgaard.dk

**Comment by Ddfulton** [comment]
Can we reopen this issue? I've been working on this for days

**Comment by clockwisemusic** [comment]
I am having this issue too, and also having the issue with 1.9.8, if we could re-open it that would be great. @Ddfulton have you made any progress?

**Comment by Ddfulton** [comment]
@clockwisemusic Nah, I asked it here http://stackoverflow.com/questions/37647994/heroku-flask-selenium-cant-find-file-in-traceback-weird-valueerror but it was put on hold for a few days. It's back up, might be worth following.

**Comment by felix185** [comment]
I am having the same issue. I am using the latest versions of Selenium and PhantomJS. When I use firefox as browser everything works. If somebody has found a solution please let me know.

**Comment by saravanakumarvelayutham** [comment]
I had this same error but the send keys has string that has white spaces , after stripping them it worked for me

**Comment by Ddfulton** [comment]
I just switched to Capybara. It works incredibly well and has a built in WebDriverWait().

**Comment by jyxt** [comment]
any update on this, this is still broken

**Comment by mrajashree** [comment]
Same error for 1.9.8

**Comment by stephanschielke** [comment]
same with phantomjs 2.1.1 in combination with selenium 3.4.3
Delays don't work for me

**Comment by jesg** [comment]
i'm leaving this issue closed.  the root cause is the content security policy on github.  this has been fixed in ghostdriver 2.0.0.
